segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/SystemJournal.java

```diff
@@ -331,37 +490,118 @@ private long applyLatestSnapshot(MetadataTransaction txn, HashMap<String, Long>
+            val chunkMetadata = (ChunkMetadata) txn.get(chunkName).get();
+            Preconditions.checkState(null != chunkMetadata);
+            chunkName = chunkMetadata.getNextChunk();
+        }
+    }
+
+    private void validateSystemSnapshot(SystemSnapshotRecord systemSnapshot) {
+        if (null != systemSnapshot) {
+            for (val segmentSnapshot : systemSnapshot.getSegmentSnapshotRecords()) {
+                segmentSnapshot.segmentMetadata.checkInvariants();
+                Preconditions.checkState(segmentSnapshot.segmentMetadata.isStorageSystemSegment());
+                Preconditions.checkState(segmentSnapshot.segmentMetadata.getChunkCount() == segmentSnapshot.chunkMetadataCollection.size());
+                if (segmentSnapshot.chunkMetadataCollection.size() == 0) {
+                    Preconditions.checkState(segmentSnapshot.segmentMetadata.getFirstChunk() == null);
+                    Preconditions.checkState(segmentSnapshot.segmentMetadata.getLastChunk() == null);
+                } else if (segmentSnapshot.chunkMetadataCollection.size() == 1) {
+                    Preconditions.checkState(segmentSnapshot.segmentMetadata.getFirstChunk() != null);
+                    Preconditions.checkState(segmentSnapshot.segmentMetadata.getFirstChunk().equals(segmentSnapshot.segmentMetadata.getLastChunk()));
+                } else {
+                    Preconditions.checkState(segmentSnapshot.segmentMetadata.getFirstChunk() != null);
+                    Preconditions.checkState(segmentSnapshot.segmentMetadata.getLastChunk() != null);
+                    Preconditions.checkState(!segmentSnapshot.segmentMetadata.getFirstChunk().equals(segmentSnapshot.segmentMetadata.getLastChunk()));
+                }
+                ChunkMetadata previous = null;
+                for (val metadata : segmentSnapshot.getChunkMetadataCollection()) {
+                    if (previous != null) {
+                        Preconditions.checkState(previous.getNextChunk().equals(metadata.getName()));
+                    }
+                    previous = metadata;
+                }
+            }
+        }
     }
 
     /**
      * Read contents from file.
      */
-    private byte[] getContents(String snapshotFile) throws ExecutionException, InterruptedException {
-        val info = chunkStorage.getInfo(snapshotFile).get();
-        val h = ChunkHandle.readHandle(snapshotFile);
-        byte[] contents = new byte[Math.toIntExact(info.getLength())];
-        long fromOffset = 0;
-        int remaining = contents.length;
-        while (remaining > 0) {
-            int bytesRead = chunkStorage.read(h, fromOffset, remaining, contents, Math.toIntExact(fromOffset)).get();
-            remaining -= bytesRead;
-            fromOffset += bytesRead;
+    private byte[] getContents(String snapshotFile) throws Exception {
+        Exception e = null;
+        for (int i = 0; i < config.getMaxJournalReadAttempts(); i++) {
```

Comment:
This is an excellent candidate for using `Retry` from `common`. It will help you eliminate all the boilerplate code with retrying and then throwing the final exception.

Reply:
This is now async